### ISSUE TITLE
Chart tool doesn't understand commas in numbers

### ISSUE THREAD

**Issue by chrisodva** [NuCivic/dkan#1183]
When creating a column chart DKAN is misinterpreting the value of numbers due to the commas in the number. File was created by saving Excel spreadsheet as a MS-DOS CSV, all defaults left as is, which adds the comma thousands marker to the number. See example at http://data.nucivic.com/dataset/chart-example. Preview the data set and take note of the value of Utah and Texas in the first column of numbers, then preview a column chart and see that Utah at 540,000 is represented on the chart as larger than TX, which is 190,000,000, apparently because DKAN is only evaluating the digits to the left of leftmost comma. (so it sees 540 and larger than 190)

Using a csv file without commas works fine.

**Comment by ajacinto** [comment]
Using other delimiters don't seem to work either. No chart would be rendered if we used say a '+' instead of a comma.

**Comment by jvitello** [comment]
I tested this again on the dev version of DKAN, and I am able to reproduce all of the issues. There are a few parts to this:

### 1) Alternative delimiters are not working
When resource files are created, there are options for `+`, `|`, and `;`, but none of these work. These alternatives would be a good workaround for using data that includes commas. We're logging this as a bug, and hopefully we'll get a resolution soon.

### 2) Commas in data points are escaped for grid previews but not for graph previews or charts
There isn't one standard way of constructing a delimited file where the data points include the delimiter character. This is what the data actually looks like in the example originally posted:

`TX,"190,383,377.00","125,739,335.00","64,644,042.00","79,335,700.00","56,740,300.00","22,378,800.00"`
`UT,"534,322.00","369,247.00","165,075.00","247,800.00","148,100.00","99,700.00"`

This is the solution implemented in Excel when exporting a csv file with this issue. We're adding an improvement ticket to our backlog to add the logic needed for identifying this special format, and getting it to work in graph previews and charts.
Not sure when we will be able to address this. The higher priority would be to fix the other delimiters, and those can be used as a workaround in most cases.

### 3) Commas in numeric data aren't interpreted by graph previews or charts
Being able to interpret...

`TX,"190,383,377.00","125,739,335.00"`

... as ...

TX | 190,383,377.00 | 125,739,335.00
--- | --- | ---

... is not the same as interpreting `190,383,377.00` as `190383377.00`. Even if we are able to interpret the csv files that use quotes as a way to separate data points, it's likely the graph previews and charts still won't be able to interpret the data.
There isn't an obvious strategy for whether or not we should work to support formatted numbers in the data, but we'll add the improvement to our backlog for consideration. There are two different use cases that we need to consider:
- Do we prioritize general user needs, and make it easier to add data in different formats and interpret it with the visualizations?
- Do we prioritize data standards and make data processing or APIs easier to use by restricting the use of formatting in data files?

I'm not sure what the DKAN team will decide, but it's helpful to get this type of feedback when we consider the best path forward.

**Comment by jvitello** [comment]
I did a little more testing with different permutations of data files, and it looks like the second issue above does work as expected. Data files like this work for graph previews and charts:

```
"Paired Label 1, Paired Label 2",2011,2012,2013
"apple, orange",0.1,0.2,0.3
"hammer,wrench",0.2,0.4,0.6
```

Since that kind of file can be interpreted correctly, it seems the commas in the numerical data are the only issue.

**Comment by ajacinto** [comment]
@jvitello - It's really just to improve the data's readability upon display. How about if the file gets uploaded in DKAN without the commas, and it's only when it's rendered as a spreadsheet or chart that the setting/option to have the thousands separator is provided?

**Comment by jvitello** [comment]
Good idea. I'll add that feature request to our queue, formatting for Grid previews.

**Comment by jvitello** [comment]
The alternative delimiters are going to be resolved in DKAN 1.13. There aren't any plans for fixing it for 1.12.

**Comment by janette** [comment]
fixed! https://github.com/NuCivic/visualization_entity/pull/66
